libsolidity/ast/AST.h

```diff
@@ -730,7 +730,7 @@ class EnumValue: public Declaration
  * User defined value types, i.e., custom types, for example, `type MyInt is int`. Allows creating a
  * zero cost abstraction over value type with stricter type requirements.
  */
-class UserDefinedValueTypeDefinition: public Declaration
+class UserDefinedValueTypeDefinition: public Declaration, public ScopeOpener
```

Comment:
I don't think the hack is worth it, especially as `canonicalName` won't help with two user defined value types defined in different source files at the top level.

Reply:
Also it's weird that this requires ScopeOpener and not Scopable.

Reply:
Why not just change this stuff? That's the only use of ``currentCanonicalName``, right? And anything that has ``TypeDeclarationAnnotation`` annotations, is a ``Declaration``? Then just remove ``currentCanonicalName`` and change 
```
	if (auto* annotation = dynamic_cast<TypeDeclarationAnnotation*>(&_node.annotation()))
		annotation->canonicalName = currentCanonicalName();
```
to sth like
```
	if (auto* annotation = dynamic_cast<TypeDeclarationAnnotation*>(&_node.annotation()))
    {
		annotation->canonicalName = dynamic_cast<Declaration&>(_node).name();
        // traverse and prepend parent scope names...
    }
```

But I only peeked at the links - might not work for reasons I didn't see.
But we should try to avoid making an AST node that doesn't open a scope a ``ScopeOpener``.

Reply:
Yes, this should be properly fixed.